src/rate_limiter.rs

```diff
@@ -0,0 +1,224 @@
+                Ok(Request(GetAccountInfo { .. })) => {
+                    if part_count > 1 {
+                        return Err(RoutingError::InvalidMessage);
+                    }
+                    CLIENT_GET_CHARGE
+                }
+                Ok(Request(PutIData { .. })) |
+                Ok(Request(PutMData { .. })) |
+                Ok(Request(MutateMDataEntries { .. })) |
+                Ok(Request(SetMDataUserPermissions { .. })) |
+                Ok(Request(DelMDataUserPermissions { .. })) |
+                Ok(Request(ChangeMDataOwner { .. })) |
+                Ok(Request(InsAuthKey { .. })) |
+                Ok(Request(DelAuthKey { .. })) => {
+                    if part_count > 1 {
+                        return Err(RoutingError::InvalidMessage);
+                    }
+                    payload.len() as u64
+                }
+                Ok(Request(Refresh(..))) |
+                Ok(Response(_)) => return Err(RoutingError::InvalidMessage),
+                Err(SerialisationError::DeserialiseExtraBytes) => {
+                    return Err(RoutingError::InvalidMessage);
+                }
+                Err(_) => {
+                    if part_count == 1 {
+                        return Err(RoutingError::InvalidMessage);
+                    }
+                    payload.len() as u64
+                }
+            }
+        } else {
+            payload.len() as u64
+        };
+
+        if bytes_to_add > allowance {
+            return Err(RoutingError::ExceedsRateLimit);
+        }
+
+        let _ = self.used.insert(*client_ip, used + bytes_to_add);
+        Ok(())
+    }
+
+    fn update(&mut self) {
+        // If there's nothing else to update, set the timestamp and return.
+        if self.used.is_empty() {
+            self.last_updated = Instant::now();
+            return;
+        }
+
+        // If the current used total has had time to fully leak away, just clear `used` and return.
+        let now = Instant::now();
+        let leak_time = (now - self.last_updated).as_secs() as f64 +
+                        ((now - self.last_updated).subsec_nanos() as f64 / 1_000_000_000.0);
+        self.last_updated = now;
+        let mut leaked_units = (RATE * leak_time) as u64;
+        if self.used.values().sum::<u64>() <= leaked_units {
+            self.used.clear();
+            return;
+        }
+
+        // Sort entries by least-used to most-used and leak each client's quota. For any client
+        // which doesn't need its full quota, the unused portion is equally distributed amongst the
+        // others.
+        let leaking_client_count = self.used.len();
+        let mut quota = leaked_units / leaking_client_count as u64;
+        let mut entries: Vec<(u64, IpAddr)> = self.used
+            .iter()
+            .map(|(ip_addr, used)| (*used, *ip_addr))
+            .collect();
+        entries.sort();
+        self.used.clear();
```

Comment:
Optional if you prefer: Can be done in the same step and safer as you cannot forget to call clear later. `let mut entries = mem::replace(&mut self.used, Default::default()).into_iter().map(...).collect();` Also a single iteration instead of two (clear() will iterate to deallocate again)